config/fcos/v1_5_exp/translate.go

```diff
@@ -348,3 +354,67 @@ func (c Config) processPackages(options common.TranslateOptions) (types.Config,
 	ts.AddFromCommonSource(yamlPath, path.New("json", "storage"), ret.Storage)
 	return ret, ts, r
 }
+
+func (c Config) handleUserGrubCfg(options common.TranslateOptions) (types.Config, translate.TranslationSet, report.Report) {
+	rendered := types.Config{}
+	ts := translate.NewTranslationSet("yaml", "json")
+	var r report.Report
+	yamlPath := path.New("yaml", "grub")
+	if len(c.Grub.Users) == 0 {
+		// No users
+		return rendered, ts, r
+	}
+
+	// create boot filesystem
+	// bootFilesystem :=
+
+	rendered.Storage.Filesystems = append(rendered.Storage.Filesystems,
+		types.Filesystem{
+			Device: "/dev/disk/by-label/boot",
+			Format: util.StrToPtr("ext4"),
+			Path:   util.StrToPtr("/boot"),
+		})
+
+	userCfgContent := []byte(buildGrubConfig(c.Grub))
+	src, compression, err := baseutil.MakeDataURL(userCfgContent, nil, !options.NoResourceAutoCompression)
+	if err != nil {
+		r.AddOnError(yamlPath, err)
+		return rendered, ts, r
+	}
+
+	// Create user.cfg file and add it to rendered config
+	rendered.Storage.Files = append(rendered.Storage.Files,
+		types.File{
+			Node: types.Node{
+				Path: "/boot/grub2/user.cfg",
+			},
+			FileEmbedded1: types.FileEmbedded1{
+				Contents: types.Resource{
+					Source:      util.StrToPtr(src),
+					Compression: compression,
+				},
+			},
+		})
+
+	ts.AddFromCommonSource(yamlPath, path.New("json", "storage"), rendered.Storage)
+	return rendered, ts, r
+}
+
+func buildGrubConfig(gb Grub) string {
+	// Process super users and corresponding passwords
+	allUsers := []string{}
+	contents := []string{"# Generated by Butane\n\n", ""}
+
+	for _, user := range gb.Users {
+		// We have already validated that user.Name is non-empty
+		allUsers = append(allUsers, user.Name)
+		if util.NotEmpty(user.PasswordHash) {
+			// Command for setting users password
+			contents = append(contents, fmt.Sprintf("password_pbkdf2 %s %s\n", user.Name, *user.PasswordHash))
+		}
+	}
+	superUserCmd := fmt.Sprintf("set superusers=\"%s\"\n", strings.Join(allUsers, " "))
+	// Slice will always be of atleast size 2 since we initialized it as such
+	contents[1] = superUserCmd
```

Comment:
This works but is a bit trickier than necessary.  Instead of `contents`, I'd be inclined to have a slice of password entries, and only build up the contents at the end of the function.